compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs

```diff
@@ -0,0 +1,192 @@
+-- Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+{-# LANGUAGE TemplateHaskell #-}
+
+-- | Constraint generator for DAML LF static verification
+module DA.Daml.LF.Verify.Generate
+  ( genPackages
+  , genChoice -- Added export to suppress the unused warning.
+  ) where
+
+import Control.Lens hiding (Context)
+import Control.Monad.Error.Class (MonadError (..))
+import Data.HashMap.Strict (singleton)
+import Control.Monad.Reader
+import qualified Data.NameMap as NM
+
+import DA.Daml.LF.Ast
+import DA.Daml.LF.Verify.Context
+import DA.Daml.LF.Verify.Subst
+import DA.Daml.LF.TypeChecker.Error
+
+data GenOutput = GenOutput
+  { _goExp :: Expr
+    -- ^ The expression, evaluated as far as possible.
+  , _goUpd :: UpdateSet
+    -- ^ The updates, performed by this expression.
+  , _goDel :: Delta
+    -- ^ The context extension, made by this expression.
+  }
+
+makeLenses ''GenOutput
+
+-- | Extend a generator output with the updates and the context extensions of a
+-- second generator output. Note that the end result will contain the first
+-- expression.
+combineGO :: GenOutput -> GenOutput
+          -- ^ The two generator outputs to combine.
+          -> GenOutput
+combineGO genOut1 genOut2
+  = extendGOUpds (genOut2 ^. goUpd)
+  $ extendGODelta (genOut2 ^. goDel)
+    genOut1
+
+updateGOExpr :: Expr
+             -- ^ The new output expression.
+             -> GenOutput
+             -- ^ The current generator output.
+             -> GenOutput
+updateGOExpr expr = set goExp expr
+
+extendGOUpds :: UpdateSet
+             -- ^ The extension of the update set.
+             -> GenOutput
+             -- ^ The current generator output.
+             -> GenOutput
+extendGOUpds upds = over goUpd (concatUpdateSet upds)
+
+extendGODelta :: Delta
+              -- ^ The additional context extension.
+              -> GenOutput
+              -- ^ The current generator output.
+              -> GenOutput
+extendGODelta delta = over goDel (concatDelta delta)
+
+genPackages :: MonadDelta m => [(PackageId, (Package, Maybe PackageName))]
+            -> m Delta
+genPackages inp = ask >>= \del0 -> foldM genPackage' del0 inp
+  where
+    -- TODO: These prime definitions are bit silly. Merge with the regular
+    -- functions.
+    -- TODO: This >>= return . concatDelta returns a lot. Figure out a way to
+    -- abstract over it. Same for this returning foldM.
+    genPackage' :: MonadDelta m => Delta
+                -> (PackageId, (Package, Maybe PackageName)) -> m Delta
+    genPackage' deli (idi, (paci, _)) =
+      -- TODO: This is getting quite unreadable.
+      foldM (\deli' modi -> genModule' deli' (PRImport idi, modi))
```

Comment:
I’m probably missing something but it looks to me like this only works if modules are sorted topologically. Otherwise, you might not be able to resolve a cross-module reference since you haven’t yet included the delta for that module?